lib/hackmode/realmode.js

```diff
@@ -0,0 +1,1159 @@
+          const nullB64 = Buffer.from(JSON.stringify(null), 'utf8').toString('base64')
+          this._getDhtRef(dnaAddress).post(Dht.DhtEvent.dataFetchResponse(e.msgId, nullB64))
+          break
+        }
+        // Create DHT data item
+        let dhtData = mem.get(entryAddress)
+        dhtData.entryAddress = entryAddress
+        dhtData.dnaAddress = dnaAddress
+        dhtData.dataFetchProviderAgentId = null
+        // Pick first agent who tracks this DNA as the provider
+        // WARNING non-deterministic if many agents track the same DNA
+        for (const [agentId, dnas] of this._ipcDnaByAgent) {
+          if (dnas.has(dnaAddress)) {
+            dhtData.dataFetchProviderAgentId = agentId
+            break
+          }
+        }
+        if (!dhtData.dataFetchProviderAgentId) {
+          log.e(this.me() + ' AgentId not found for DNA ' + dnaAddress + ' ; during handleDhtEvent(dataFetch)')
+        }
+        log.t(this.me() + ' dataFetchResponse:', dhtData)
+        dhtData = Buffer.from(JSON.stringify(dhtData), 'utf8').toString('base64')
+        this._getDhtRef(dnaAddress).post(Dht.DhtEvent.dataFetchResponse(e.msgId, dhtData))
+        break
+      default:
+        throw new Error('unhandled dht event type ' + e.type + ' ' + JSON.stringify(e))
+    }
+  }
+
+  /**
+   * send/publish a p2p message to a Peer
+   * @private
+   */
+  async _p2pSend (peerAddress, obj) {
+    // log.t(this.me() + ' >>> ' + this.nick(peerAddress) + ' - ' + obj.type)
+    return this._p2p.publishReliable(
+      [peerAddress],
+      msgpack.encode(obj).toString('base64')
+    )
+  }
+
+  /**
+   * send/publish a p2p message to all know Peers
+   * @private
+   */
+  async _p2pSendAll (obj) {
+    // log.t(this.me() + ' >>> ' + this._peerList + ' - ' + obj.type)
+    return this._p2p.publishReliable(
+      this._peerList,
+      msgpack.encode(obj).toString('base64')
+    )
+  }
+
+  /**
+   * _p2p is notifying me of p2p events
+   * @private
+   */
+  _handleP2pEvent (evt) {
+    // log.t(this.me() + '._handleP2pEvent()', evt.type)
+    switch (evt.type) {
+      case 'peerConnect':
+        log.t(this.me() + '.peerConnect', this.nick(evt.peerAddress))
+        // Bookkeep known peers
+        this._peerList.push(evt.peerAddress)
+        // Gossip back my tracked DNA (and agents)
+        // convert to array
+        let dnaByAgent = []
+        for (const [agentId, dnas] of this._ipcDnaByAgent) {
+          // log.t(this.me() + '.gossipAllTracks.' + agentId, Array.from(dnas))
+          for (const dnaAddress of dnas) {
+            let dnaAgent = []
+            dnaAgent.push(agentId)
+            dnaAgent.push(dnaAddress)
+            dnaByAgent.push(dnaAgent)
+          }
+        }
+        // log.t(this.me() + '.gossipAllTracks', this._ipcDnaByAgent)
+        log.t(this.me() + '.gossipAllTracks', dnaByAgent)
+        this._p2pSend(evt.peerAddress, {
+          type: 'gossipAllTracks',
+          dnaByAgent: dnaByAgent
+        })
```

Comment:
I still feel like there's something here, in that the dht code should be gossiping the peer data internally, so we shouldn't have to do this.

There *is* a bootstrapping problem... the transport layer requires the "advertise" / bootstrapping in the conductor config. And this is plugging that gap for now, and it isn't that much overhead when we can't have that many nodes in any case, so I'm cool with this for now until we figure out something better.

Reply:
I don't see how the data DHTs could gossip stuff by their own since we never give them peerTransports.
If you are talking about the p2p DHT, that could probably work if we added the 'dna tracking data' as data in the p2p DHT.

Reply:
the dht-spec relies on being able to message other peers through the "gossip" events, and on the remote side accepting posts of "remoteGossipBundle" events. If those are hooked up then yes, the data DHTs can talk to each other...

And we * do * give them peerTransports, that's what the agentId to transport (hc://) ids are, right?

Reply:
Yes, its the peerTransport field but its not "advertise" thats what I meant.
I think I still need some verbal explanation on how this could work :)